File: site/11-Backtracking-and-Bit-Manipulation/python/n_queens.py
def solve_n_queens(n):
    board = [['.' for _ in range(n)] for _ in range(n)]
    
    def is_safe(row, col):
        # Check column
        for i in range(row):
            if board[i][col] == 'Q':
                return False
        # Check main diagonal
        i, j = row - 1, col - 1
        while i >= 0 and j >= 0:
            if board[i][j] == 'Q':
                return False
            i -= 1
            j -= 1
        # Check anti diagonal
        i, j = row - 1, col + 1
        while i >= 0 and j < n:
            if board[i][j] == 'Q':
                return False
            i -= 1
            j += 1
        return True

    def backtrack(row):
        if row == n:
            return True
        for col in range(n):
            if is_safe(row, col):
                board[row][col] = 'Q'
                if backtrack(row + 1):
                    return True
                board[row][col] = '.' # Backtrack
        return False

    if backtrack(0):
        return board
    return None

if __name__ == "__main__":
    n = 4
    solution = solve_n_queens(n)
    if solution:
        print(f"Found a solution for {n}-Queens:")
        for row in solution:
            print(" ".join(row))
    else:
        print(f"No solution found for {n}-Queens.")

File: 11-Backtracking-and-Bit-Manipulation/cpp/n_queens.cpp
#include <iostream>
#include <vector>
#include <string>

bool isSafe(const std::vector<std::string>& board, int row, int col, int n) {
    // Check column
    for (int i = 0; i < row; ++i) {
        if (board[i][col] == 'Q') return false;
    }
    // Check upper left diagonal
    for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; --i, --j) {
        if (board[i][j] == 'Q') return false;
    }
    // Check upper right diagonal
    for (int i = row - 1, j = col + 1; i >= 0 && j < n; --i, ++j) {
        if (board[i][j] == 'Q') return false;
    }
    return true;
}

bool backtrack(std::vector<std::string>& board, int row, int n) {
    if (row == n) return true;
    for (int col = 0; col < n; ++col) {
        if (isSafe(board, row, col, n)) {
            board[row][col] = 'Q';
            if (backtrack(board, row + 1, n)) return true;
            board[row][col] = '.'; // Backtrack
        }
    }
    return false;
}

int main() {
    int n = 4;
    std::vector<std::string> board(n, std::string(n, '.'));
    
    if (backtrack(board, 0, n)) {
        std::cout << "Found a solution for " << n << "-Queens:\n";
        for (const auto& row : board) {
            for (char c : row) {
                std::cout << c << " ";
            }
            std::cout << "\n";
        }
    } else {
        std::cout << "No solution found.\n";
    }
    
    return 0;
}
